boa_interner/src/lib.rs

```diff
@@ -67,44 +67,161 @@
+}
+
+impl<'a> From<&'a [u16]> for JStrRef<'a> {
+    fn from(s: &'a [u16]) -> Self {
+        JStrRef::Utf16(s)
+    }
+}
+
+impl<'a, const N: usize> From<&'a [u16; N]> for JStrRef<'a> {
+    fn from(s: &'a [u16; N]) -> Self {
+        JStrRef::Utf16(s)
+    }
+}
+
+/// A double reference to an interned string inside [`Interner`].
+///
+/// [`JSInternedStrRef::utf8`] returns an [`Option`], since
+/// not every `UTF-16` string is fully representable as a
+/// `UTF-8` string (because of unpaired surrogates). However,
+/// every `UTF-8` string is representable as a `UTF-16` string,
+/// so [`JSInternedStrRef::utf8`] returns a
+/// [<code>&\[u16\]</code>][std::slice].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct JSInternedStrRef<'a, 'b> {
+    utf8: Option<&'a str>,
+    utf16: &'b [u16],
+}
+
+impl<'a, 'b> JSInternedStrRef<'a, 'b> {
+    /// Returns the inner reference to the interned string in
+    /// `UTF-8` encoding. if the string is not representable in
+    /// `UTF-8`, returns [`None`]
+    pub fn utf8(&self) -> Option<&'a str> {
+        self.utf8
+    }
+
+    /// Returns the inner reference to the interned string in
+    /// `UTF-16` encoding.
+    pub fn utf16(&self) -> &'b [u16] {
+        self.utf16
+    }
+
+    /// Joins the result of both possible strings into a common
+    /// type.
+    ///
+    /// If `self` is representable by a `UTF-8` string and the
+    /// `prioritize_utf8` argument is set, it will prioritize calling `f`,
+    /// and will only call `g` if `self` is only representable by a
+    /// `UTF-16` string. Otherwise, it will directly call
+    /// `g`.
+    pub fn join<F, G, T>(self, f: F, g: G, prioritize_utf8: bool) -> T
+    where
+        F: FnOnce(&'a str) -> T,
+        G: FnOnce(&'b [u16]) -> T,
+    {
+        if prioritize_utf8 {
+            if let Some(str) = self.utf8 {
+                return f(str);
+            }
+        }
+        g(self.utf16)
+    }
+
+    /// Same as [`join`][`JSInternedStrRef::join`], but where you can pass an
+    /// additional context.
+    ///
+    /// Useful when you have a `&mut Context` context that cannot
+    /// be borrowed by both closures at the same time.
+    pub fn join_with_context<C, F, G, T>(self, f: F, g: G, ctx: C, prioritize_utf8: bool) -> T
+    where
+        F: FnOnce(&'a str, C) -> T,
+        G: FnOnce(&'b [u16], C) -> T,
+    {
+        if prioritize_utf8 {
+            if let Some(str) = self.utf8 {
+                return f(str, ctx);
+            }
+        }
+        g(self.utf16, ctx)
+    }
+
+    /// Converts both string types into a common type `C`.
+    ///
+    /// If `self` is representable by a `UTF-8` string and the
+    /// `prioritize_utf8` argument is set, it will prioritize converting
+    /// its `UTF-8` representation first,
+    /// and will only convert its `UTF-16` representation if it is only representable by a
+    /// `UTF-16` string. Otherwise, it will directly convert its `UTF-16`
+    /// representation.
```

Comment:
Each line should be up to 100 characters long, but as long as possible. This seems a bit chaotic. Seems to happen in all documentation comments in this file.

Reply:
Yeah, let me readjust the lengths